aggregate.go

```diff
@@ -90,32 +144,55 @@ func (a *Aggregator) Aggregate(bundles []Bundle) ([][]Bundle, error) {
 	return aggregatedBundles, nil
 }
 
-func (a *Aggregator) aggregatedBundlesMissVersionBundle(bundles, newGroup []Bundle) bool {
-	// delta is the number of version bundles diverging compared to the aggregated
-	// list of version bundles making up a release. In case delta is greater than
-	// 0, the release misses delta version bundles. In case delta is lower than 0,
-	// the release has delta version bundles too much. The latter should be way
-	// more improbable to happen.
-	var delta int
-	{
-		desireCount := distinctCount(bundles)
-		currentCount := len(newGroup)
-
-		delta = desireCount - currentCount
+func (a *Aggregator) walkTreeAndAddLeaves(n *node, bundles []Bundle) {
+	if len(n.leaves) == 0 {
+		for _, b := range bundles {
+			n.leaves = append(n.leaves, &node{bundle: b})
+		}
+		return
+	}
+
+	for _, leaf := range n.leaves {
+		a.walkTreeAndAddLeaves(leaf, bundles)
 	}
+}
 
-	if delta != 0 {
-		v, err := aggregateReleaseVersion(newGroup)
-		if err != nil {
-			a.logger.Log("level", "error", "message", "failed aggregating release version", "stack", fmt.Sprintf("%#v", err))
-		} else {
-			a.logger.Log("level", "debug", "message", fmt.Sprintf("release misses %d version bundles", delta), "version", v)
+func (a *Aggregator) walkTreeAndAggregate(n *node, bundles []Bundle) [][]Bundle {
+	// If current node is leaf, then return bundle if there are no conflicts.
+	if len(n.leaves) == 0 {
+		// Only aggregate bundle groups that don't have conflicting dependencies.
+		for i, b1 := range bundles {
+			for j, b2 := range bundles {
+				// No need to self-verify.
+				if i == j {
+					continue
+				}
+
+				if a.bundlesConflictWithDependencies(b1, b2) {
+					return [][]Bundle{}
+				}
+
+				if a.bundlesConflictWithDependencies(b2, b1) {
+					return [][]Bundle{}
+				}
```

Comment:
Does it have to be called in both directions? Can't bundlesConflictWithDependencies be changed to do it instead?

Reply:
This is how it used to be. I'll check if there are any consequences in making the check both ways. That would make sense.